### ISSUE TITLE
Heatmap with InterpretDL

### ISSUE THREAD

**Issue by Cho-Hong-Seok** [ultralytics/ultralytics#10155]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I want to detect fire and smoke objects with the YOLOv8,9 model and then display the detection results as a heatmap. I want to utilize InterpretDL, but I don't know how to implement it. Can you advise me on how to implement it?
I would also like to know how to utilize Grad-cam.

Thank you!

> InterpretDL Github 
> https://github.com/PaddlePaddle/InterpretDL

**Comment by glenn-jocher** [comment]
Hey there! 😊

It's great you're exploring advanced visualization techniques like heatmaps with YOLOv8,9 models. While directly integrating InterpretDL with Ultralytics YOLO models isn't straightforward due to framework differences, you can still achieve your goal by following these general steps:

1. **Detection with YOLO:** First, run your fire and smoke detection using YOLOv8,9 and save the results. This can easily be done with the predict mode provided by the Ultralytics API.

```python
from ultralytics import YOLO

# Load model
model = YOLO('yolov8.pt')  # tweak based on your specific model

# Detect and save results
results = model('path/to/your/image.jpg')
results.save()  # Saves detected images in 'runs/detect' by default
```

2. **Heatmap Visualization:** For integrating with InterpretDL or Grad-CAM implementations, you'll need the raw model outputs. This generally involves using the detected regions along with the model's backbone. Since InterpretDL is designed for PaddlePaddle models, and YOLOv8,9 models are in PyTorch, you'd primarily look at similar PyTorch alternatives like [`pytorch-gradcam`](https://github.com/jacobgil/pytorch-grad-cam).

Here's a simplified example using Grad-CAM with PyTorch (assuming you have a way to deal with framework differences or find a tool similar to InterpretDL that supports PyTorch):

```python
from pytorch_grad_cam import GradCAM
from pytorch_grad_cam.utils.image import show_cam_on_image

# Assuming `model` and `target_layers` are defined
cam = GradCAM(model=model, target_layers=target_layers, use_cuda=True)

# Generate heatmap
target_category = None  # Define if specific class; otherwise, highest scoring class
grayscale_cam = cam(input_tensor=image_tensor, target_category=target_category)

# Overlay heatmap on original image
heatmap = show_cam_on_image(np.array(image) / 255., grayscale_cam)
```

Remember, for the actual implementation, you'll need to adapt this to your specific case, especially handling the conversion of detections to a format compatible with heatmap generation tools.

Hope this helps you get started! The community is here if you have more questions. Good luck with your fire and smoke detection project! 🔥💨

**Comment by sizhongyibanhts** [comment]
https://github.com/z1069614715/objectdetection_script/tree/master
<img width="890" alt="image" src="https://github.com/ultralytics/ultralytics/assets/44115021/92e1a3de-b9bf-49da-9204-a907d26d9618">

**Comment by Cho-Hong-Seok** [comment]
> https://github.com/z1069614715/objectdetection_script/tree/master <img alt="image" width="890" src="https://private-user-images.githubusercontent.com/44115021/325908140-92e1a3de-b9bf-49da-9204-a907d26d9618.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQxMjE1NzIsIm5iZiI6MTcxNDEyMTI3MiwicGF0aCI6Ii80NDExNTAyMS8zMjU5MDgxNDAtOTJlMWEzZGUtYjliZi00OWRhLTkyMDQtYTkwN2QyNmQ5NjE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI2VDA4NDc1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmNjViZWIxOWI5MzA4MGE3N2I0MzdkZTdlZmVhMjQzNWIxNTFjYjhkZmY2ZDkzN2Q4ZjBlYzBiNTliNTA0OGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.51ufnvy4Dg1cVRarxS4Pk6mYq7TTdPJIrgm0iz6HhjY">

Thanks for the answer! 
I currently have fire and smoke data in my project, bboxed and segmented with YOLOv9 and YOLOv8 models.   I saw the grad-cam material on the github address you mentioned, but it's not easy to follow due to my lack of skills. The guide video is also in Chinese and the resolution is low, so it's hard to understand. Do you have an ipynb file like a tutorial?

**Comment by glenn-jocher** [comment]
Hey there! 😊

It sounds like you're doing fantastic work with fire and smoke detection using YOLOv8 and YOLOv9 models! I'm sorry to hear the Grad-CAM guide wasn't easy to follow. Unfortunately, we don't have an ipynb tutorial ready for Grad-CAM integration at the moment.

However, integrating Grad-CAM with YOLO models essentially involves extracting the last convolutional layer of the model and applying Grad-CAM techniques on it. While we work on providing more accessible materials, here's a very basic snippet that shows how you might approach this with a PyTorch model:

```python
from pytorch_grad_cam import GradCAM
from pytorch_grad_cam.utils.image import show_cam_on_image
import cv2
import numpy as np

# Assuming `model` is your YOLO model and `target_layer` is your last conv layer
cam = GradCAM(model=model, target_layers=[target_layer], use_cuda=False)

# Prepare your input image
# image should be np.array of shape (height, width, 3)
input_tensor = preprocess_image(image, mean=[0,0,0], std=[1,1,1], resize=True)

# Get the cam mask
grayscale_cam = cam(input_tensor=input_tensor, targets=None)
grayscale_cam = grayscale_cam[0, :]

# In this line "image" should be in the range [0, 1] and of shape (height, width, 3)
visualization = show_cam_on_image(image, grayscale_cam, use_rgb=True)

# Now visualize or save your result
cv2.imshow('Grad-CAM', visualization)
cv2.waitKey(0)
```

This example presupposes that you have the `pytorch_grad_cam` library installed and your model and image ready. "preprocess_image" would be a function you define based on your model's input requirements.

If you're looking for more step-by-step guidance or implementations, it might help to look into resources specific to PyTorch's handling of Grad-CAM or, if using another framework, their respective methods.

Keep up the fantastic work, and I hope this snippet gives you a starting point. Feel free to reach out if you have more questions! 🚀

**Comment by sizhongyibanhts** [comment]
Thank you, but I have one more question. When training large language models, such as chatglm and llama, have you ever thought about visualizing the gradient results? We now continue pretraining (pretrain) In the process, it is difficult to evaluate what the model has learned. You can see that in my picture about target detection yesterday, there is a heat map. Through the heat map, I can clearly judge the quality of different labeling methods. Similarly, in the training of large language models, it may be more helpful to understand the training process by seeing the attention scores of the model for input and output at each step. I wonder if you have any practical experience in this area? To evaluate the quality of a model, you cannot just use the recall score, but you should look at what the model has learned.谢谢你，但是我还有一个问题。在训练大语言模型，例如chatglm和llama的时候，您有没有想过可视化梯度结果？我们现在进行继续预训练（pretrain）的过程中，很难评估模型学会了什么。您可以看到，在我昨天的关于目标检测的图片中，有一张热力图，通过热力图我可以清晰的判断出不同标注方式的好坏。同理，在大语言模型的训练中，也许看到模型对于输入和每一步输出的注意力分数，对理解训练过程更有帮助。不知道您在这方面有没有可用的实践经验呢？评估模型好坏，不能只用recall分数，而是应该去看看模型到底学会了什么。My colleagues are using tools like mmdetection and paddleocr and I will recommend your tool to them.我的同事们正在使用mmdetection和paddleocr之类的工具，我会向他们推荐一下您的工具的。发自我的荣耀手机-------- 原始邮件 --------发件人： Glenn Jocher ***@***.***>日期： 2024年4月27日周六 00:58收件人： ultralytics/ultralytics ***@***.***>抄送： Han Tianshuo ***@***.***>, Comment ***@***.***>主    题： Re: [ultralytics/ultralytics] Heatmap with InterpretDL (Issue #10155)
Hey there! 😊
It sounds like you're doing fantastic work with fire and smoke detection using YOLOv8 and YOLOv9 models! I'm sorry to hear the Grad-CAM guide wasn't easy to follow. Unfortunately, we don't have an ipynb tutorial ready for Grad-CAM integration at the moment.
However, integrating Grad-CAM with YOLO models essentially involves extracting the last convolutional layer of the model and applying Grad-CAM techniques on it. While we work on providing more accessible materials, here's a very basic snippet that shows how you might approach this with a PyTorch model:
from pytorch_grad_cam import GradCAM
from pytorch_grad_cam.utils.image import show_cam_on_image
import cv2
import numpy as np

# Assuming `model` is your YOLO model and `target_layer` is your last conv layer
cam = GradCAM(model=model, target_layers=[target_layer], use_cuda=False)

# Prepare your input image
# image should be np.array of shape (height, width, 3)
input_tensor = preprocess_image(image, mean=[0,0,0], std=[1,1,1], resize=True)

# Get the cam mask
grayscale_cam = cam(input_tensor=input_tensor, targets=None)
grayscale_cam = grayscale_cam[0, :]

# In this line "image" should be in the range [0, 1] and of shape (height, width, 3)
visualization = show_cam_on_image(image, grayscale_cam, use_rgb=True)

# Now visualize or save your result
cv2.imshow('Grad-CAM', visualization)
cv2.waitKey(0)
This example presupposes that you have the pytorch_grad_cam library installed and your model and image ready. "preprocess_image" would be a function you define based on your model's input requirements.
If you're looking for more step-by-step guidance or implementations, it might help to look into resources specific to PyTorch's handling of Grad-CAM or, if using another framework, their respective methods.
Keep up the fantastic work, and I hope this snippet gives you a starting point. Feel free to reach out if you have more questions! 🚀

**Comment by sizhongyibanhts** [comment]
Download the corresponding python file in this picture, install the relevant packages with pip, and then modify some minor problems (I only encountered a problem with the numpy data type). You said you were doing fire and smoke detection. I found that the shape of the flame is very complex. We have made relevant annotations before, but we have never evaluated the annotation method through heat maps. Flame can be very easy to identify because his features are obvious. But how do you differentiate between flames and car lights at night? For smoke data, it is more necessary to use heat maps for evaluation. In our data set, there is a lot of smoke formed from water vapor. How to prevent misidentification is very important. If you have any news, you can share it with me. We also have many interesting scenes here. For example, do you have any opinions on small target recognition and whether personnel wear safety helmets correctly (in compliance with wearing regulations)? The video website you see is equivalent to China’s YouTube. There is a lot of content in it. It would be very interesting if it could be translated.下载这张图片里面对应的python文件，pip 安装相关的包，然后修改一些小问题（我只遇到了一个numpy数据类型的问题）。你说你在进行火和烟雾检测。我发现火焰形状非常复杂，之前我们也进行过相关标注，但是一直没有通过热力图的方式进行过标注方式的评估。火焰可能非常好识别，因为他的特征很明显。但是如何区分夜间的火焰与车灯？对于烟雾数据，更需要利用热力图进行评估。在我们的数据集中，有很多水蒸气形成的烟雾。如何防止误识别很重要。有消息可以和我一起分享，我们这里也有很多有意思的场景。例如，对于小目标识别，以及人员是否正确佩戴安全帽（是否符合佩戴规范），您有什么意见吗？你看到的视频网站，相当于中国的youtube，里面有很多内容，如果能翻译过来，一定很有意思。发自我的荣耀手机-------- 原始邮件 --------发件人： Hongseok Cho ***@***.***>日期： 2024年4月26日周五 16:53收件人： ultralytics/ultralytics ***@***.***>抄送： Han Tianshuo ***@***.***>, Comment ***@***.***>主    题： Re: [ultralytics/ultralytics] Heatmap with InterpretDL (Issue #10155)

https://github.com/z1069614715/objectdetection_script/tree/master 

Thanks for the answer!
I currently have fire and smoke data in my project, bboxed and segmented with YOLOv9 and YOLOv8 models.   I saw the grad-cam material on the github address you mentioned, but it's not easy to follow due to my lack of skills. The guide video is also in Chinese and the resolution is low, so it's hard to understand. Do you have an ipynb file like a tutorial?

**Comment by sizhongyibanhts** [comment]
Hello, acting on intuition, I have a very unconventional idea. Suppose you have a 10*10 piece of smoke. You can first mark the entire smoke, and then mark a 5*5 area inside the smoke to enhance the model's understanding of the smoke. Or you can rotate and stretch these areas (by stretching, I mean twisting a square image into a trapezoid, so that the shape of the smoke also changes). These methods can enhance the model's understanding of smoke characteristics. I think it is possible to make the model more practical, although the evaluation scores (e.g. recall map) will be very low.We have been engaged in related work for several months, why no one has ever cared about what the model has learned. China's state-owned enterprises may be a little impetuous.Regarding the target recognition task, I still hope to communicate more with you and learn more usage skills.-------- 原始邮件 --------发件人： Hongseok Cho ***@***.***>日期： 2024年4月26日周五 16:53收件人： ultralytics/ultralytics ***@***.***>抄送： Han Tianshuo ***@***.***>, Comment ***@***.***>主    题： Re: [ultralytics/ultralytics] Heatmap with InterpretDL (Issue #10155)

https://github.com/z1069614715/objectdetection_script/tree/master 

Thanks for the answer!
I currently have fire and smoke data in my project, bboxed and segmented with YOLOv9 and YOLOv8 models.   I saw the grad-cam material on the github address you mentioned, but it's not easy to follow due to my lack of skills. The guide video is also in Chinese and the resolution is low, so it's hard to understand. Do you have an ipynb file like a tutorial?

**Comment by glenn-jocher** [comment]
Hey there! 🙌

Your unconventional idea sounds truly interesting and innovative! Marking a smoke area and then a specific part inside it can indeed offer the model a nuanced understanding of what constitutes smoke. This is a creative approach to data annotation that could improve model performance in recognizing complex patterns, especially in diverse conditions.

Rotation and stretching (or skewing) annotations, as you've described, could potentially increase the robustness of the model against various shapes and orientations of smoke. It's a nice strategy to simulate different perspectives and smoke dynamics without needing an extensive dataset covering all possible scenarios.

Regarding the concern about what the model has learned, it's indeed crucial but often overlooked due to a focus on performance metrics. However, understanding the underlying patterns the model perceives is key to refining and deploying it effectively, especially in safety-critical tasks like smoke detection.

For further enhancing model understanding, you might explore visualization techniques like the mentioned Grad-CAM or attention maps for deep dives into models' focus areas during training and inference.

I'm glad you're looking forward to more communication and learning! Sharing experiences and insights could really help in navigating these challenges. Although I don't have an ipynb ready as a tutorial, your idea and approach surely prompt a good discussion on improving practical applicability beyond traditional metrics. 

Let's keep the conversation going and learn from each other's experiences. Stay innovative! 👍

**Comment by Cho-Hong-Seok** [comment]
> 안녕하세요, 직감에 따라 매우 파격적인 아이디어가 있습니다. 10*10 연기가 있다고 가정합니다. 먼저 전체 연기를 표시한 다음 연기 내부에 5*5 영역을 표시하여 연기에 대한 모델의 이해를 높일 수 있습니다. 또는 이 영역을 회전하고 늘릴 수 있습니다(늘리면 정사각형 이미지를 사다리꼴로 비틀어 연기의 모양도 변경됨). 이러한 방법은 연기 특성에 대한 모델의 이해를 향상시킬 수 있습니다. 모델을 더 실용적으로 만드는 것이 가능하다고 생각하지만 평가 점수(예: 회상 맵)는 매우 낮을 것입니다. 우리는 몇 달 동안 관련 작업에 종사해 왔는데, 왜 아무도 모델이 배운 것에 대해 신경 쓰지 않았습니다. 중국의 국영 기업은 약간 성급할 수 있습니다. 대상 인식 과제에 관해서는 여전히 여러분과 더 많이 소통하고 더 많은 사용법을 배우고 싶습니다.-------- 原始邮件 --------发件人: Hongseok Cho ***@***.***>日期: 2024年4月26日周五 16:53收件人: ultralytics/ultralytics ***@***.***>抄送: Han Tianshuo ***@***.***>, Comment ***@***.***>主 题: Re: [ultralytics/ultralytics] Heatmap with InterpretDL (Issue #10155) https://github.com/z1069614715/objectdetection_script/tree/master 답변 해 주셔서 감사합니다! 현재 내 프로젝트에는 YOLOv9 및 YOLOv8 모델로 bboxed 및 분할 된 화재 및 연기 데이터가 있습니다. 언급 한 github 주소에서 grad-cam 자료를 보았지만 기술이 부족하여 따라하기가 쉽지 않습니다. 가이드 동영상도 중국어로 해상도가 낮기 때문에 이해하기 어렵습니다. 튜토리얼과 같은 ipynb 파일이 있습니까? - 이 이메일에 직접 회신하거나, GitHub에서 보거나, 구독을 취소합니다. 댓글을 달았기 때문에 받는 것입니다. 메시지 ID: ***@***.***>

Thanks for the answer!
Let me explain the error that happened to me.
I ran yolov9_heatmap.py.
Here is the error I got.

File "yolov9_heatmap.py", line 191, in <module>
    model = yolov9_heatmap(**get_params())
  File "yolov9_heatmap.py", line 109, in __init__
    colors = np.random.uniform(0, 255, size=(len(model_names), 3)).astype(np.int)
  File "/home/dilab03/anaconda3/envs/YOLOv6/lib/python3.8/site-packages/numpy/__init__.py", line 305, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'int'.
`np.int` was a deprecated alias for the builtin `int`. To avoid this error in existing code, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations`ile "yolov9_heatmap.py", line 191, in <module>
    model = yolov9_heatmap(**get_params())
  File "yolov9_heatmap.py", line 109, in __init__
    colors = np.random.uniform(0, 255, size=(len(model_names), 3)).astype(np.int)
  File "/home/dilab03/anaconda3/envs/YOLOv6/lib/python3.8/site-packages/numpy/__init__.py", line 305, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'int'.
`np.int` was a deprecated alias for the builtin `int`. To avoid this error in existing code, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations

Chatgpt에게 오류를 물어보았고 수정한 결과 아래와 같은 오류가 또 발생했습니다. 
yolov9-c summary: 724 layers, 50970180 parameters, 0 gradients, 237.7 GFLOPs
Traceback (most recent call last):
  File "yolov9_heatmap.py", line 193, in <module>
    model(r'./Fault-Detection-6/test/images', 'result')
  File "yolov9_heatmap.py", line 173, in __call__
    self.process(f'{img_path}/{img_path_}', f'{save_path}/{img_path_}')
  File "yolov9_heatmap.py", line 144, in process
    grayscale_cam = self.method(tensor, [self.target])
  File "/home/dilab03/anaconda3/envs/YOLOv6/lib/python3.8/site-packages/pytorch_grad_cam/base_cam.py", line 188, in __call__
    return self.forward(input_tensor,
  File "/home/dilab03/anaconda3/envs/YOLOv6/lib/python3.8/site-packages/pytorch_grad_cam/base_cam.py", line 74, in forward
    outputs = self.activations_and_grads(input_tensor)
  File "yolov9_heatmap.py", line 67, in __call__
    post_result, pre_post_boxes, post_boxes = self.post_process(model_output[0])
  File "yolov9_heatmap.py", line 57, in post_process
    logits_ = result[:, 4:]
TypeError: list indices must be integers or slices, not tuple

이후 더 이상 진전이 없습니다. 도와주실 수 있을까요?

**Comment by glenn-jocher** [comment]
안녕하세요! 😊

먼저, 기발한 아이디어와 질문에 대해 공유해주셔서 감사합니다. NumPy 관련 오류를 해결하신 것을 축하드립니다. 두 번째 오류는 모델 출력이 예상대로 작동하지 않아 발생하는 것으로 보입니다. `model_output[0]`이 리스트를 반환하는 것 같은데, `result[:, 4:]`를 사용하기 위해서는 NumPy array나 PyTorch tensor 같은 형태여야 합니다.

다음은 오류를 수정할 수 있는 간략한 해결 방안입니다:

`model_output[0]`의 타입을 확인하고, 필요하다면 NumPy array나 PyTorch tensor로 변환해주세요. 예를 들어, `model_output[0]`이 리스트라면, 다음과 같이 수정할 수 있습니다.

```python
...
logits_ = np.array(result)[:, 4:]  # NumPy array로 변환
# 또는 PyTorch tensor의 경우:
# logits_ = torch.tensor(result)[:, 4:]
...
```

이 조정으로 모델 출력을 올바르게 처리할 수 있을 겁니다. 🔥

연기 인식에 대해 말씀하셨던 그 초혁신적인 아이디어는 정말 흥미롭습니다. 모델이 연기를 더 잘 이해할 수 있도록 돕는 독창적인 접근 방식이 매력적이네요. 밤에 화염과 차량 불빛을 구분하는 문제는 매우 도전적인데, 지적하신 대로, 이러한 복잡한 시나리오에서 높은 수준의 세밀한 주석과 훈련 데이터의 다양성이 중요할 것 같습니다.

안전 헬멧 착용과 같은 특정 조건을 충족하는 상황에서의 소규모 객체 인식과 같은 다양한 시나리오에 대한 깊이 있는 논의와 공유가 기대됩니다. 🚒👷

어떤 진전이 있으시면 언제든지 나누어 주세요. 서로 배우고 성장할 수 있는 기회가 되길 바랍니다!

**Comment by sizhongyibanhts** [comment]
我没用过yolov9 今天我试一下吧 I have never used yolov9, I will try it today.发自我的荣耀手机-------- 原始邮件 --------发件人： Hongseok Cho ***@***.***>日期： 2024年4月28日周日 01:01收件人： ultralytics/ultralytics ***@***.***>抄送： Han Tianshuo ***@***.***>, Comment ***@***.***>主    题： Re: [ultralytics/ultralytics] Heatmap with InterpretDL (Issue #10155)

안녕하세요, 직감에 따라 매우 파격적인 아이디어가 있습니다. 1010 연기가 있다고 가정합니다. 먼저 전체 연기를 표시한 다음 연기 내부에 55 영역을 표시하여 연기에 대한 모델의 이해를 높일 수 있습니다. 또는 이 영역을 회전하고 늘릴 수 있습니다(늘리면 정사각형 이미지를 사다리꼴로 비틀어 연기의 모양도 변경됨). 이러한 방법은 연기 특성에 대한 모델의 이해를 향상시킬 수 있습니다. 모델을 더 실용적으로 만드는 것이 가능하다고 생각하지만 평가 점수(예: 회상 맵)는 매우 낮을 것입니다. 우리는 몇 달 동안 관련 작업에 종사해 왔는데, 왜 아무도 모델이 배운 것에 대해 신경 쓰지 않았습니다. 중국의 국영 기업은 약간 성급할 수 있습니다. 대상 인식 과제에 관해서는 여전히 여러분과 더 많이 소통하고 더 많은 사용법을 배우고 싶습니다.-------- 原始邮件 --------发件人: Hongseok Cho @.>日期: 2024年4月26日周五 16:53收件人: ultralytics/ultralytics @.>抄送: Han Tianshuo @.>, Comment @.>主 题: Re: [ultralytics/ultralytics] Heatmap with InterpretDL (Issue #10155) https://github.com/z1069614715/objectdetection_script/tree/master 답변 해 주셔서 감사합니다! 현재 내 프로젝트에는 YOLOv9 및 YOLOv8 모델로 bboxed 및 분할 된 화재 및 연기 데이터가 있습니다. 언급 한 github 주소에서 grad-cam 자료를 보았지만 기술이 부족하여 따라하기가 쉽지 않습니다. 가이드 동영상도 중국어로 해상도가 낮기 때문에 이해하기 어렵습니다. 튜토리얼과 같은 ipynb 파일이 있습니까? - 이 이메일에 직접 회신하거나, GitHub에서 보거나, 구독을 취소합니다. 댓글을 달았기 때문에 받는 것입니다. 메시지 ID: @.***>

Thanks for the answer!
Let me explain the error that happened to me.
I ran yolov9_heatmap.py.
Here is the error I got.
File "yolov9_heatmap.py", line 191, in 
model = yolov9_heatmap(**get_params())
File "yolov9_heatmap.py", line 109, in init
colors = np.random.uniform(0, 255, size=(len(model_names), 3)).astype(np.int)
File "/home/dilab03/anaconda3/envs/YOLOv6/lib/python3.8/site-packages/numpy/init.py", line 305, in getattr
raise AttributeError(former_attrs[attr])
AttributeError: module 'numpy' has no attribute 'int'.
np.int was a deprecated alias for the builtin int. To avoid this error in existing code, use int by itself. Doing this will not modify any behavior and is safe. When replacing np.int, you may wish to use e.g. np.int64 or np.int32 to specify the precision. If you wish to review your current use, check the release note link for additional information.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations`ile "yolov9_heatmap.py", line 191, in 
model = yolov9_heatmap(**get_params())
File "yolov9_heatmap.py", line 109, in init
colors = np.random.uniform(0, 255, size=(len(model_names), 3)).astype(np.int)
File "/home/dilab03/anaconda3/envs/YOLOv6/lib/python3.8/site-packages/numpy/init.py", line 305, in getattr
raise AttributeError(former_attrs[attr])
AttributeError: module 'numpy' has no attribute 'int'.
np.int was a deprecated alias for the builtin int. To avoid this error in existing code, use int by itself. Doing this will not modify any behavior and is safe. When replacing np.int, you may wish to use e.g. np.int64 or np.int32 to specify the precision. If you wish to review your current use, check the release note link for additional information.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
Chatgpt에게 오류를 물어보았고 수정한 결과 아래와 같은 오류가 또 발생했습니다.
yolov9-c summary: 724 layers, 50970180 parameters, 0 gradients, 237.7 GFLOPs
Traceback (most recent call last):
File "yolov9_heatmap.py", line 193, in 
model(r'./Fault-Detection-6/test/images', 'result')
File "yolov9_heatmap.py", line 173, in call
self.process(f'{img_path}/{img_path_}', f'{save_path}/{img_path_}')
File "yolov9_heatmap.py", line 144, in process
grayscale_cam = self.method(tensor, [self.target])
File "/home/dilab03/anaconda3/envs/YOLOv6/lib/python3.8/site-packages/pytorch_grad_cam/base_cam.py", line 188, in call
return self.forward(input_tensor,
File "/home/dilab03/anaconda3/envs/YOLOv6/lib/python3.8/site-packages/pytorch_grad_cam/base_cam.py", line 74, in forward
outputs = self.activations_and_grads(input_tensor)
File "yolov9_heatmap.py", line 67, in call
post_result, pre_post_boxes, post_boxes = self.post_process(model_output[0])
File "yolov9_heatmap.py", line 57, in post_process
logits_ = result[:, 4:]
TypeError: list indices must be integers or slices, not tuple
이후 더 이상 진전이 없습니다. 도와주실 수 있을까요?

**Comment by glenn-jocher** [comment]
@sizhongyibanhts 안녕하세요! 😊 YOLOv9를 시도해보시려는 것을 들으니 정말 기쁩니다. 여러분이 겪고 계신 오류에 대해 조금 도움을 드리고자 합니다.

첫 번째 오류는 NumPy 데이터 타입 관련 이슈로, 이미 해결하신 것 같네요! `np.int` 대신 `int`, `np.int64`, 또는 `np.int32`를 사용하신 것이 정확한 접근 방법입니다. 👍

두 번째 오류는 모델의 출력 처리 중 문제가 발생한 것 같습니다. 오류 메시지에 따르면, `result`가 리스트로 예상되는 위치에 튜플이나 정수가 사용되어야하는 상황에서 오류가 발생했습니다. 

대부분의 경우, PyTorch 모델은 `torch.Tensor` 객체를 반환합니다. `result[: , 4:]`의 사용은, `result`가 NumPy 배열이나 PyTorch 텐서일 때 유효합니다. 만약 `result`가 리스트라면, 먼저 적절한 데이터 타입으로 변환해야 합니다.

아래는 이런 상황을 처리하기 위한 예시 코드입니다:

```python
# result가 list 형태일 경우, NumPy 배열로 변환
result = np.array(model_output[0])
logits_ = result[:, 4:]
```

만약 `model_output[0]`이 이미 텐서라면, `result` 변환 없이 직접 사용하셔도 됩니다.

어떠한 데이터 타입 문제든, 중요한 것은 대상 변수의 데이터 타입을 명확히 파악하고, 해당 타입에 맞는 올바른 인덱싱이나 조작을 수행하는 것입니다. 👩‍💻

추가 질문이 있으시면 언제든지 문의주세요! YOLOv9 탐험에 행운을 빌어요! 🚀

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #25548](https://github.com/ultralytics/ultralytics/pull/25548), titled **“Replace predict feature-map dumps with class activation heatmaps.”**

Key changes include:

- Replaced raw layer-by-layer feature-map dumps with practical PyTorch class activation heatmaps using LayerCAM.
- Updated `visualize=True` during prediction to generate heatmap overlays alongside prediction outputs.
- Applied confidence and class filters, with support for visualizing a configurable number of top predictions.
- Limited activation-map generation to supported PyTorch workflows, excluding tensor sources, augmentation, and embedding workflows.
- Updated inference and documentation to reflect the new heatmap behavior.

These changes should provide more interpretable visual explanations by highlighting image regions that influence predictions, while ensuring visualizations focus on relevant detections.

The changes were implemented by [@Y-T-G](https://github.com/Y-T-G), with contributions from [@glenn-jocher](https://github.com/glenn-jocher).

To test the latest changes, install directly from the `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release.

Please test this update and share whether it resolves the issue. Thank you 🙏 for reporting it, and any further feedback is welcome if the problem persists.

**Comment by glenn-jocher** [comment]
@Cho-Hong-Seok, the improvement contributed by @Y-T-G in #25548 is now merged. You can use `visualize=True` during prediction to create a class-score LayerCAM overlay for fire/smoke detections without integrating InterpretDL, and `classes=[...]` limits the explanation to selected classes.
